lib/keys.js

```diff
@@ -33,63 +51,103 @@ class Keys {
     this.rest.requestJson(options, (e, r, json) => {
       if (e) {
         logger.error(e);
-        callback(e);
+        if (callback) callback(e);
         return;
       }
 
-      callback(e, json);
+      // Refill the cache.
+      this.keyCache = {};
+      json.forEach((key) => {
+        this.keyCache[key.name] = key;
+      });
+
+      if (callback) callback(e, json);
     });
   }
 
   delete(name, callback) {
     const options = {
       method: 'delete',
-      uri: `/api/3/sshkeys/${this.username}/${name}/`,
+      uri: `/api/3/sshkeys/${this.username}/${name}`,
     };
 
     this.rest.request(options, (e) => {
       if (e) {
-        callback(e);
+        if (callback) callback(e);
         return;
       }
 
-      callback();
+      // Update the cache
+      if (this.keyCache !== null) delete this.keyCache[name];
+
+      if (callback) callback();
     }).end();
   }
 
-  add(obj, callback) {
+  save(obj, callback) {
     const options = {
-      method: 'post',
-      uri: `/api/3/sshkeys/${this.username}/`,
+      method: 'put',
+      uri: `/api/3/sshkeys/${this.username}/${obj.name}`,
       json: obj,
     };
 
     this.rest.requestJson(options, (e, r, json) => {
       if (e) {
         logger.error(e);
-        callback(e);
+        if (callback) callback(e);
         return;
       }
 
-      callback(e, json);
+      // Update the cache
+      if (this.keyCache === null) this.keyCache = {};
+      this.keyCache[json.name] = json;
+
+      if (callback) callback(e, json);
     });
   }
 
   update(name, obj, callback) {
     const options = {
       method: 'patch',
-      uri: `/api/3/sshkeys/${this.username}/${name}/`,
+      uri: `/api/3/sshkeys/${this.username}/${name}`,
       json: obj,
     };
 
     this.rest.requestJson(options, (e, r, json) => {
       if (e) {
         logger.error(e);
-        callback(e);
+        if (callback) callback(e);
+        return;
+      }
+
+      // Update the cache
+      if (this.keyCache === null) this.keyCache = {};
+      this.keyCache[json.name] = json;
+
+      if (callback) callback(e, json);
+    });
+  }
+
+  find(givenKey, callback) {
+    this.list((e, keys) => {
+      if (e) {
+        if (callback) callback(e);
         return;
       }
 
-      callback(e, json);
+      for (let i = 0; i < keys.length; i++) {
```

Comment:
Couldn't this be a `forEach`?

Reply:
Not the same, I am returning once I find the key. In `forEach()` return just advances to the next item.